### PR TITLE
[keccak] Theta gate circuit

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### Problem & Solution

We separate the theta step config here and test it with the test cases acquired at another branch.

Note: should merge after #82

### CODE REVIEW

**Review by miha-stopar — COMMENTED** [review]
Nice! I only have one keccak newbie question - why are there + and ^ theta function versions in keccak_arith.rs and plain.rs? Test in keccah_arith.rs uses ^ version, while test in theta.rs uses + version - any reason for this?

File: keccak256/src/gates/theta.rs [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            (0..5)
+                .cartesian_product(0..5)
+                .map(|(x, y)| {
+                    let new_state =
+                        meta.query_advice(state[5 * x + y], Rotation::next());
+                    let old_state =
+                        meta.query_advice(state[5 * x + y], Rotation::cur());
+                    let right = old_state
+                        + column_sum[(x + 4) % 5].clone()
+                        + Expression::Constant(F::from(13))
```
Comment (CPerezz) [inline keccak256/src/gates/theta.rs:51]: When we hardcode numbers in the impl. Would be nice to add a comment or some info on "Where does this cone from".

In this case would be the check of the base 13 IIUC. But would be nice to add it.

**Review by CPerezz — APPROVED** [review]
LGTM!

Nice job @ChihChengLiang :smile:

File: keccak256/src/gates/theta.rs (hunk 1 above)
Comment (ChihChengLiang) [reply keccak256/src/gates/theta.rs:51]: Nice catch, I will fix it.

**Comment by ChihChengLiang** [comment]
> Nice! I only have one keccak newbie question - why are there + and ^ theta function versions in keccak_arith.rs and plain.rs? Test in keccah_arith.rs uses ^ version, while test in theta.rs uses + version - any reason for this?

@miha-stopar 
The `plain.rs` is the usual keccak implementation and our circuit will compute keccak the way `keccak_arith.rs` does.
The `keccak_arith` uses tricks to turn xor(^) operations into simple arithmetic addition(+) operations to save table lookups.

For example, if we want to XOR two 4 bit values `a = 0101` and `b = 1111`. We can transform them into base 3 numbers

- `a3 = 0 * 3^3 + 1 * 3^2 + 0 * 3^1 + 1 * 3^0`
- `b3 = 1 * 3^3 + 1 * 3^2 + 1 * 3^1 + 1 * 3^0`

The sum of them would be `c3 = 1 * 3^3 + 2 * 3^2 + 1 * 3^1 + 2 * 3^0`, with coefficients `1, 2, 1, 2`
And since we know

|x|y|x^y|x+y|
|-|-|-|-|
|0|0|0|0|
|0|1|1|1|
|1|0|1|1|
|1|1|0|2|

(The table works because f(x+y) ->g(x^y) is injective)

We can map the coefficients back to the XORed result of individual bits `1, 0, 1, 0`, thus getting the `a^b = 1010`.

The trick works if we know exactly how many XOR we have to perform in the theta step.

**Comment by miha-stopar** [comment]
Nice! Thanks for a detailed explanation @ChihChengLiang !!

**Review by miha-stopar — APPROVED** [review]

**Review by han0110 — APPROVED** [review]
LGTM! Nice work!

**Comment by ChihChengLiang** [comment]
Thank you @miha-stopar @CPerezz and @han0110 for the review!

**Pull request merged**